Add source map spec tests

takikawa opened the pull request:
This is a draft PR for issue #504.

The PR adds integration to run the in-development source map specification tests. The tests are accessed via a git submodule, though the current location of the tests is likely not their final official location. The tests are still under development as well.

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,91 @@
+  return json;
+}
+
+// Known failures due to intentional implementation choices or due to bugs.
+const skippedTests = [
+  // Versions are explicitly checked a bit loosely.
+  "versionNumericString",
+  // Stricter sources array checking isn't implemented.
+  "sourcesNotStringOrNull",
+  "sourcesAndSourcesContentBothNull",
+  // Stricter names array checking isn't implemented.
+  "namesMissing",
+  "namesNotString",
+  // This check isn't as strict in this library.
+  "invalidMappingNotAString1",
+  // A mapping segment with no fields is technically invalid in the spec.
+  "invalidMappingSegmentWithZeroFields",
+  // These tests fail due to imprecision in the spec about the 32-bit limit.
+  "invalidMappingSegmentWithColumnExceeding32Bits",
+  "invalidMappingSegmentWithOriginalLineExceeding32Bits",
+  "invalidMappingSegmentWithOriginalColumnExceeding32Bits",
+  // A large VLQ that should parse, but currently does not.
+  "validMappingLargeVLQ",
+  // The library currently doesn't check the types of offset lines/columns.
+  "indexMapOffsetLineWrongType",
+  "indexMapOffsetColumnWrongType",
+  // The spec is not totally clear about this case.
+  "indexMapInvalidBaseMappings",
+  // The spec's definition of overlap can be refined
+  "indexMapInvalidOverlap",
+  // Not clear if this test makes sense, but spec isn't clear on behavior
+  "validMappingNullSources"
+]
```
Are these skipped tests specific to the mozilla sourcemap implementation
```suggestion
// Known failures due to intentional implementation choices or due to bugs.
const skippedTests = [
  // Versions are explicitly checked a bit loosely.
  "versionNumericString",
  // Stricter sources array checking isn't implemented.
  "sourcesNotStringOrNull",
  "sourcesAndSourcesContentBothNull",
  // Stricter names array checking isn't implemented.
  "namesMissing",
  "namesNotString",
  // This check isn't as strict in this library.
  "invalidMappingNotAString1",
  // A mapping segment with no fields is technically invalid in the spec.
  "invalidMappingSegmentWithZeroFields",
  // These tests fail due to imprecision in the spec about the 32-bit limit.
  "invalidMappingSegmentWithColumnExceeding32Bits",
  "invalidMappingSegmentWithOriginalLineExceeding32Bits",
  "invalidMappingSegmentWithOriginalColumnExceeding32Bits",
  // A large VLQ that should parse, but currently does not.
  "validMappingLargeVLQ",
  // The library currently doesn't check the types of offset lines/columns.
  "indexMapOffsetLineWrongType",
  "indexMapOffsetColumnWrongType",
  // The spec is not totally clear about this case.
  "indexMapInvalidBaseMappings",
  // The spec's definition of overlap can be refined
  "indexMapInvalidOverlap",
  // Not clear if this test makes sense, but spec isn't clear on behavior
  "validMappingNullSources"
]
```

takikawa replied:
Yes, for example the "versionNumericString" test is skipped because this library specifically accepts the case where the version field is a number in a string. Other implementations have different behavior (e.g., the [source map validator](https://github.com/jkup/source-map-validator) uses this library but layers more strict checking on top).

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,91 @@
+  exports[`test from source map spec tests, name: ${testCase.name}`] =
+    async function (assert) {
+      const json = await readJSON(`./source-map-tests/resources/${testCase.sourceMapFile}`);
+      let sourceMapFailed = false;
```
This does not seem to be used!

takikawa replied:
Thanks for catching that, I think it was a vestige of an earlier attempt. I'll remove it next time I update the patch.

bomsy commented:
Thanks for this patch. this is looking great. I just had a quick look. I'll be testing it in a little bit

takikawa commented:
Note: I updated the patch with support for checking generated->original mapping (before it just checked original->generated). Now it depends on PR #507 as well.

I also updated it with checking for transitive mapping. Both of these are pretty localized changes.

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,129 @@
+        map.eachMapping(() => {});
+        map.destroy();
+      } catch (exn) {
+        if (testCase.sourceMapIsValid)
```
```suggestion
        if (testCase.sourceMapShouldBeValid)
```

bomsy replied:
This is not related directly to this patch but it just a tiny suggestion while trying the understanding the code.
`sourceMapShouldBeValid` might make it clearer what the expectation is.

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,129 @@
+      if (!testCase.sourceMapIsValid)
+        assert.fail("Expected invalid source map but loaded successfully");
+      if (testCase.testActions) {
+        for (let testAction of testCase.testActions) {
```
```suggestion
        for (const testAction of testCase.testActions) {
```

bomsy replied:
Maybe we could use `const` since `testAction` is not reassigned.

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,129 @@
+      column: action.generatedColumn,
+    });
+
+    for (let intermediateMapPath of action.intermediateMaps) {
```
```suggestion
    for (const intermediateMapPath of action.intermediateMaps) {
```

bomsy replied:
Since `intermediateMapPath` is not reassigned

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,129 @@
+  return SourceMapConsumer.with(rawSourceMap, null, (consumer) => {
+    let mappedPosition = consumer.generatedPositionFor({
+      source: action.originalSource,
+      line: action.originalLine + 1,
```
```suggestion
      line: mapLine(action.originalLine),
```
```
function mapLine(line) {
  return line + 1;
}
```

bomsy replied:
So we seem to be adding `1` to all the lines in most places here, we can have a util function which adds the line.

bomsy commented:
Thanks @takikawa. I tested out the patch and it works nicely. I've added a few inline comments.

Once the tests have been moved to the main repo we can updated and land this patch. Do you know the timeline for this?

Also could we automate the update of the git sub-module by  adding the command to the npm task for running tests in the `package.json`
This simple task seems to work for me. 
```
"test": "git submodule update --init --recursive; node test/run-tests.js",
```
This is a basic inital solution, it can be improved with followups later

takikawa commented:
> Once the tests have been moved to the main repo we can updated and land this patch. Do you know the timeline for this?

Thanks for the comments! We've now imported the tests into a more official repo (https://github.com/tc39/source-map-tests). I'll work on updating this PR to address your comments and update for the new location soon (hopefully this week).

takikawa commented:
I've update the PR and pointed it at the new repo, and addressed your comments (except one, commented above). Happy to address any further feedback.

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,208 @@
+ * http://opensource.org/licenses/BSD-3-Clause
+ */
+
+const fs = require("node:fs/promises");
```
```suggestion
const fs = require("fs");
```

bomsy replied:
This is likely causing the failure in the jobs. Can you fix.

bomsy commented on test/test-spec-tests.js:
```diff
@@ -0,0 +1,208 @@
+ * http://opensource.org/licenses/BSD-3-Clause
+ */
+
+const fs = require("node:fs/promises");
```
```suggestion
const fs = require("fs").promises;
```

bomsy replied:
So we try this to fix the failing job?

bomsy commented:
@takikawa  The patch  LGTM. Added a comment to resolve the failure in the jobs.

bomsy reviewed (approved):
LGTM! Lets land this Thanks @takikawa for the patch.

bomsy commented:
@takikawa Applying the fix to see if the 

> @takikawa The patch LGTM. Added a comment to resolve the failure in the jobs.

Fixed the failure in the jobs, so we can land the patch.
